src/app-layer-ssl.c

```diff
@@ -250,54 +255,416 @@ static void SSLSetTxDetectFlags(void *vtx, uint8_t dir, uint64_t flags)
+                                         const uint32_t input_len)
+{
+    uint8_t *initial_input = input;
+
+    if (!(HAS_SPACE(SSLV3_CLIENT_HELLO_RANDOM_LEN))) {
+        SCLogDebug("TLS handshake invalid length");
+        SSLSetEvent(ssl_state,
+                    TLS_DECODER_EVENT_HANDSHAKE_INVALID_LENGTH);
+        return -1;
+    }
+
+    /* Skip random */
     input += SSLV3_CLIENT_HELLO_RANDOM_LEN;
 
+    return (input - initial_input);
+}
+
+static inline int TLSDecodeHSHelloSessionID(SSLState *ssl_state,
+                                            uint8_t *input,
+                                            const uint32_t input_len)
+{
+    uint8_t *initial_input = input;
+
     if (!(HAS_SPACE(1)))
         goto invalid_length;
 
-    /* skip session id */
-    uint8_t session_id_length = *(input++);
+    uint8_t session_id_length = *input;
+    input += 1;
+
     if (session_id_length != 0) {
         ssl_state->flags |= SSL_AL_FLAG_SSL_CLIENT_SESSION_ID;
     }
 
+    if (!(HAS_SPACE(session_id_length)))
+        goto invalid_length;
+
     input += session_id_length;
 
+    return (input - initial_input);
+
+invalid_length:
+    SCLogDebug("TLS handshake invalid length");
+    SSLSetEvent(ssl_state,
+                TLS_DECODER_EVENT_HANDSHAKE_INVALID_LENGTH);
+    return -1;
+}
+
+static inline int TLSDecodeHSHelloCipherSuites(SSLState *ssl_state,
+                                               uint8_t *input,
+                                               const uint32_t input_len)
+{
+    uint8_t *initial_input = input;
+
     if (!(HAS_SPACE(2)))
         goto invalid_length;
 
-    /* skip cipher suites */
-    uint16_t cipher_suites_length = input[0] << 8 | input[1];
+    uint16_t cipher_suites_length = *input << 8 | *(input + 1);
     input += 2;
 
-    input += cipher_suites_length;
+    if (!(HAS_SPACE(cipher_suites_length)))
+        goto invalid_length;
+
+    if ((ssl_state->current_flags & SSL_AL_FLAG_STATE_CLIENT_HELLO) &&
+            ssl_config.enable_ja3) {
+        int rc;
+
+        JA3Buffer *ja3_cipher_suites = Ja3BufferInit();
+        if (ja3_cipher_suites == NULL)
+            return -1;
+
+        uint16_t processed_len = 0;
+        while (processed_len < cipher_suites_length)
+        {
+            if (!(HAS_SPACE(2))) {
+                Ja3BufferFree(ja3_cipher_suites);
+                goto invalid_length;
+            }
+
+            uint16_t cipher_suite = *input << 8 | *(input + 1);
+            input += 2;
+
+            if (TLSDecodeValueIsGREASE(cipher_suite) != 1) {
+                rc = Ja3BufferAddValue(ja3_cipher_suites, cipher_suite);
+                if (rc != 0) {
+                    Ja3BufferFree(ja3_cipher_suites);
+                    return -1;
+                }
+            }
+
+            processed_len += 2;
+        }
+
+        rc = Ja3BufferAppendBuffer(ssl_state->ja3_str, ja3_cipher_suites);
+        if (rc != 0) {
+            return -1;
```

Comment:
not all error paths seem to lead to freeing of ja3_cipher_suites?

Reply:
On error it's freed in Ja3BufferAppendBuffer, so it's cleared on all error paths.